Improve post Site Scan action items

delawski opened the pull request:
## Summary

<!-- Please reference the issue(s) this PR fixes, if one exists. For significant changes, opening an issue first for discussion is usually preferable. -->
Fixes #6689

This PR improves the action items for when there are theme and/or plugin incompatibilities.

Here are the todo items from the original issue:

- [x] The action item to review and suppress plugins should rather only appear if there are plugins with incompatibilities, and this action item should appear inside the plugins box. 
- [x] In the incompatible themes box, this is probably where the recommendation to change to Reader mode should be. Or else to consider switching to an alternative theme (see below).
- [x] In the boxes for themes and plugins, there should be links to the AMP Compatible tabs for the the Theme Install and Plugin Install screens, if theme/plugin installation is available (i.e. if filesystem is not read only). If the theme/plugin installation is not available, there should rather be `ExternalLink`s going to the theme/plugin ecosystem pages on amp-wp.org.
- [x] When a site scan completes, the `suppressed_plugins` option should be re-fetched so that the suppressed table is updated with which plugins have validation errors.
- [x] Discontinue doing synchronous re-validation of validated URLs when updating suppressed plugins. Instead of doing this, initiating a re-scan should be done.

Here's the updated Site Scan panel on the AMP Settings screen where the plugin/theme action items are located inside the respective sub-panels:

![Screenshot 2021-11-08 at 19 16 02](https://user-images.githubusercontent.com/478735/140796083-54ce486e-bae0-4434-a641-efa19f2d903c.png)

## Checklist

- [ ] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [ ] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

delawski committed:
Move post-scan action items to plugins/themes boxes

delawski committed:
Refetch plugin suppression data after completing site scan

delawski committed:
Prevent rendering empty notice element

delawski commented:
@westonruter The only item left to do is adding links to the AMP Compatible tabs for themes and plugins. I was wondering about the filesystem check you mentioned in the original issue. I think we need to not only check the filesystem but also user capabilities. Do you think we could reuse some of the logic from [`can_install_theme()`](https://github.com/ampproject/amp-wp/blob/68807c9228c9a9c9f1eba4993cfc64910972c99f/src/Admin/ReaderThemes.php#L381)? I guess we could do something similar for plugins and then pass two flags to the frontend so that appropriate links are rendered on the Settings page.

westonruter commented:
> The only item left to do is adding links to the AMP Compatible tabs for themes and plugins. I was wondering about the filesystem check you mentioned in the original issue. I think we need to not only check the filesystem but also user capabilities. Do you think we could reuse some of the logic from [`can_install_theme()`](https://github.com/ampproject/amp-wp/blob/68807c9228c9a9c9f1eba4993cfc64910972c99f/src/Admin/ReaderThemes.php#L381)? I guess we could do something similar for plugins and then pass two flags to the frontend so that appropriate links are rendered on the Settings page.

@delawski That's right. If the filesystem is not writable, the `install_themes` and `install_plugins` capabilities will return false due to logic in `map_meta_cap()`. But then going ahead with the additional checks in `\AmpProject\AmpWP\Admin\ReaderThemes::can_install_theme()` would make it even more robust, although I think it may be overkill since we don't actually need to perform the installation ourselves. I think that was done specifically for the case where a site may require SFTP credentials to be provided to do the update. In contrast, all we need to do is provide a link to the theme/plugin install screens. So for that, all we need to do is check if `current_user_can('install_themes')` and `current_user_can('install_plugins')`. The logic would be as follows:

```php
$theme_link  = current_user_can( 'install_themes' ) ? admin_url( '/theme-install.php?browse=amp-compatible' ) : 'https://amp-wp.org/ecosystem/themes/';
$plugin_link = current_user_can( 'install_plugins' ) ? admin_url( '/plugin-install.php?tab=amp-compatible' ) : 'https://amp-wp.org/ecosystem/plugins/';
```

westonruter commented:
And when the capability checks return false, we should have the external link icon to go with the links.

delawski committed:
Fix unit tests

delawski committed:
Add links to AMP compatible themes and plugins

delawski committed:
Add unit tests for `SiteScanSourcesList` component

delawski committed:
Stay in loading state until the Wizard scan has started

This prevents the progress bar starting at 100% in the Onboarding Wizard site scan.

delawski committed:
Do not mutate state outside of reducer

delawski commented:
@westonruter Thank you for the suggestions. I've used the proposed logic in fd06053:

![Screenshot 2021-11-15 at 17 46 52](https://user-images.githubusercontent.com/478735/141830304-245685df-6f74-4bb2-8b10-882e4141224e.png)

Also, I found a bug in the Site Scan in the Wizard. The progress bar has been shown too early, when the state was not stable yet causing the progress bar to start at 100%. This has been fixed in 3621298.

Lastly, I've included a Site Scan reducer one-line fix (253e5fd) to the issue you've revealed in https://github.com/ampproject/amp-wp/pull/6705#discussion_r747121045.

delawski commented:
I've just noticed a new action item on the [original issue](https://github.com/ampproject/amp-wp/issues/6689):

> Discontinue doing synchronous re-validation of validated URLs when updating suppressed plugins. Instead of doing this, initiating a re-scan should be done.

I'll take care of that next.

delawski committed:
Trigger async scan when updating suppressed plugins instead of doing synchronous validation

delawski committed:
Merge branch 'develop' into update/site-scan-results-panel

delawski commented:
> I've just noticed a new action item on the [original issue](https://github.com/ampproject/amp-wp/issues/6689):
> 
> > Discontinue doing synchronous re-validation of validated URLs when updating suppressed plugins. Instead of doing this, initiating a re-scan should be done.
> 
> I'll take care of that next.

This has been addressed in 87e609c:

https://user-images.githubusercontent.com/478735/142208618-1e292c6d-fb8b-4fa9-9fab-b9a3775a3137.mp4

westonruter commented:
Regarding this phpunit failure:

![image](https://user-images.githubusercontent.com/134745/142255581-009ed57f-969d-4650-849b-4aeb4ad4e269.png)

This seems to be an issue in PHP 8 and/or WP 5.9-alpha, so I'll investigate. It's unrelated to this PR anyway.

westonruter commented:
> This seems to be an issue in PHP 8 and/or WP 5.9-alpha, so I'll investigate. It's unrelated to this PR anyway.

It's this change in WordPress core: https://github.com/WordPress/wordpress-develop/commit/5d5c9bf6a78a52a48981abb651c321da10109fde

Easy to fix. See https://github.com/ampproject/amp-wp/pull/6727.

westonruter committed:
Merge branch 'develop' of github.com:ampproject/amp-wp into update/site-scan-results-panel

* 'develop' of github.com:ampproject/amp-wp:
  Fix test_amp_remove_paired_endpoint in WP 5.9 by removing empty query string

westonruter commented:
I just realized there is another factor for the recommendation matrix. Namely, if the user has a plugin suppressed then we should never recommend Standard mode, which we are doing now:

![image](https://user-images.githubusercontent.com/134745/142299756-b0eb946e-2792-41ac-a548-50de49937b4c.png)

If a plugin is being suppressed, then this normally means that you're in a paired AMP mode and you don't want the plugin running on the AMP version but you do want it running on the non-AMP version. If you are in Standard mode, then essentially the user should rather just have deactivated the plugin entirely as that is how it would be behaving.

So this would look something like this:

```diff
diff --git a/assets/src/components/use-template-mode-recommendation/index.js b/assets/src/components/use-template-mode-recommendation/index.js
index d6b9dee9d..dfea95155 100644
--- a/assets/src/components/use-template-mode-recommendation/index.js
+++ b/assets/src/components/use-template-mode-recommendation/index.js
@@ -60,12 +60,14 @@ export function useTemplateModeRecommendation() {
  * @param {boolean} args.hasFreshSiteScanResults Whether fresh site scan results are available.
  * @param {boolean} args.hasThemeIssues          Whether the site scan found themes with AMP incompatibility.
  * @param {boolean} args.userIsTechnical         Whether the user answered yes to the technical question.
+ * @param {boolean} args.hasSuppressedPlugins
  */
 export function getTemplateModeRecommendation( {
 	hasPluginIssues,
 	hasFreshSiteScanResults,
 	hasThemeIssues,
 	userIsTechnical,
+	hasSuppressedPlugins,
 } ) {
 	/* eslint-disable @wordpress/no-unused-vars-before-return */
 	const mobileRedirectionNote = __( 'If automatic mobile redirection is enabled, the AMP version of the content will be served on mobile devices. If AMP-to-AMP linking is enabled, once users are on an AMP page, they will continue navigating your AMP content.', 'amp' );
@@ -298,17 +300,18 @@ export function getTemplateModeRecommendation( {
 					],
 				},
 				[ TRANSITIONAL ]: {
-					recommendationLevel: NOT_RECOMMENDED,
+					recommendationLevel: hasSuppressedPlugins ? RECOMMENDED : NOT_RECOMMENDED,
 					details: [
 						transitionalModeDescription,
 						notRecommendedDueToCompleteCompatibility,
 					],
 				},
 				[ STANDARD ]: {
-					recommendationLevel: RECOMMENDED,
+					recommendationLevel: hasSuppressedPlugins ? NOT_RECOMMENDED : RECOMMENDED,
 					details: [
 						standardModeDescription,
 						__( 'Recommended as you have an AMP-compatible theme and no issues were detected with any of the plugins on your site.', 'amp' ),
+						hasSuppressedPlugins ? __( 'Not recommended because you have suppressed plugins.', 'amp' ) : null,
 					],
 				},
 			};
@@ -326,7 +329,7 @@ export function getTemplateModeRecommendation( {
 					],
 				},
 				[ TRANSITIONAL ]: {
-					recommendationLevel: NEUTRAL,
+					recommendationLevel: hasSuppressedPlugins ? RECOMMENDED : NEUTRAL,
 					details: [
 						transitionalModeDescription,
 						__( 'Recommended if you can’t commit to choosing plugins that are AMP compatible when extending your site. This mode will make it easy to keep AMP content even if non-AMP-compatible plugins are used later on.', 'amp' ),
@@ -337,6 +340,7 @@ export function getTemplateModeRecommendation( {
 					details: [
 						standardModeDescription,
 						__( 'Recommended if you can commit to always choosing plugins that are AMP compatible when extending your site.', 'amp' ),
+						hasSuppressedPlugins ? __( 'Not recommended because you have suppressed plugins.', 'amp' ) : null,
 					],
 				},
 			};
```

westonruter committed:
Improve theme/plugin site scan copy

westonruter commented on assets/src/components/site-scan-context-provider/index.js:
```diff
@@ -283,13 +291,26 @@ export function SiteScanContextProvider( {
+		if (
+			status === STATUS_READY &&
+			savedOptions?.suppressed_plugins &&
+			Object.keys( savedOptions?.suppressed_plugins ).length > 0
```
Nit:

```suggestion
			Object.keys( savedOptions.suppressed_plugins ).length > 0
```

westonruter commented on assets/src/components/site-scan-results/plugins-with-amp-incompatibility.js:
```diff
@@ -3,29 +3,38 @@
  */
 import { __ } from '@wordpress/i18n';
-import { useMemo } from '@wordpress/element';
+import { createInterpolateElement, useMemo } from '@wordpress/element';
```
Fancy.

westonruter commented on assets/src/components/site-scan-results/plugins-with-amp-incompatibility.js:
```diff
@@ -40,6 +49,26 @@ export function PluginsWithAmpIncompatibility( { className, slugs = [], ...props
+			{ showHelpText && (
+				<p>
+					{ createInterpolateElement(
+						__( 'Because of plugin issues we’ve uncovered, you may want to <a>review and suppress plugins</a>.', 'amp' ),
```
Aside: With `createInterpolateElement`, what happens when you have more than one `<a>` in your string?

delawski replied:
You can use whatever tag name you wish, e.g. you could do something like this:

```js
createInterpolateElement( 'Visit <ArchiveLink>archive</ArchiveLink> or return to <HomeLink>homepage</HomeLink>', {
	ArchiveLink: <a href="/archive" />,
	HomeLink: <a href="/" />,
} );
```

westonruter committed:
Remove obsolete SUPPRESSED_PLUGINS_ERRORING_URLS option

delawski committed:
Remove redundant optional chaining operator

Co-authored-by: Weston Ruter <westonruter@google.com>

delawski committed:
Don't recommend Standard mode if there are any suppressed plugins

delawski committed:
Add test for new prop

delawski commented:
> I just realized there is another factor for the recommendation matrix. Namely, if the user has a plugin suppressed then we should never recommend Standard mode, which we are doing now

Thanks for providing that diff. I've added it along with a logic for determining the `hasSuppressedPlugins` in [0de088d](https://github.com/ampproject/amp-wp/pull/6698/commits/0de088d32f49fb277049f375cea4acebabffa40d).

westonruter commented on assets/src/components/use-template-mode-recommendation/index.js:
```diff
@@ -31,21 +31,28 @@ export function useTemplateModeRecommendation() {
 
+		const currentOptions = { ...originalOptions, ...savedOptions };
+
+		// Plugins are considered suppressed only if they are active (included in the `suppressible_plugins` object).
+		const hasSuppressedPlugins = Object.entries( currentOptions?.suppressed_plugins || {} )
+			.some( ( [ slug, suppressed ] ) => suppressed && Boolean( currentOptions.suppressible_plugins?.[ slug ] ) );
```
Nicely done

westonruter commented on assets/src/components/options-context-provider/index.js:
```diff
@@ -55,6 +58,52 @@ export function OptionsContextProvider( { children, optionsRestPath, populateDef
+			setFetchingPluginSuppression( true );
+
+			try {
+				const fetchedPluginSuppression = await apiFetch( {
+					path: addQueryArgs( optionsRestPath, {
+						_fields: [ 'suppressed_plugins', 'suppressible_plugins' ],
+					} ),
+				} );
```
❤️

westonruter commented on assets/src/components/options-context-provider/index.js:
```diff
@@ -55,6 +58,52 @@ export function OptionsContextProvider( { children, optionsRestPath, populateDef
+					return;
+				}
+
+				setOriginalOptions( {
+					...originalOptions,
+					...fetchedPluginSuppression,
+				} );
```
This is great. I was thinking that this could have the effect of overriding a user's unsaved changes to plugin suppression, but this won't be the case the original options are updated, not the modified otpions.

westonruter reviewed (approved):
🥇
